File: openTheLock.cpp
# include<iostream>
# include <bits/stdc++.h>
using namespace std;

int findingTurns(vector<string> deadEnds, string target)
{
    unordered_set<string> dead(deadEnds.begin(), deadEnds.end());
    unordered_set<string> vis;
    queue<pair<string, int>> q;

    string start = "0000";
    if(dead.find(start) != dead.end()) return -1;
    q.push({start, 0});
    vis.insert(start);

    while(!q.empty())
    {
        // auto state, turns = q.front();
        string state = q.front().first;
        int turns = q.front().second;
        q.pop();

        if(state == target) return turns;

        for(int i = 0; i < 4; i++)
        {
            string s1 = state;
            string s2 = state;
            s1[i] = (state[i] - '0' + 1) % 10 + '0';
            s2[i] = (state[i] - '0' - 1 + 10) % 10 + '0';

            if(vis.find(s1) == vis.end() && dead.find(s1) == dead.end())
            {
                vis.insert(s1);
                q.push({s1, turns + 1});
            }
            if(vis.find(s2) == vis.end() && dead.find(s2) == dead.end())
            {
                vis.insert(s2);            
                q.push({s2, turns + 1});
            }            
            
        }
    }
    return -1;
}


int main(){
    vector<string> deadEnds = {"0201","0101","0102","1212","2002"};
    cout<<findingTurns(deadEnds, "0202");    

    return 0;
}

File: openTheLock.java
import java.util.*;

public class openTheLock {

    static class lockState {
        String state;
        int turns;

        public lockState(String state, int turns)
        {
            this.state = state;
            this.turns =turns;
        }           
    }


    public static void main(String[] args) {
        String[] deadends = {"0201", "0101", "0102", "1212", "2002"};
        String target = "0202";
        System.out.println("Minimum turns: " + findingTurns(deadends, target));
    }
    

    public static int findingTurns(String[] deadends, String target)
    {
        Set<String> dead = new HashSet<>(Arrays.asList(deadends)); 
        Set<String> vis = new HashSet<>();
        Queue<lockState> q = new LinkedList<>();

        String start = "0000";
        if(dead.contains(start)) return -1;
        vis.add(start);
        q.offer(new lockState(start, 0));

        while (!q.isEmpty()) {
            lockState curr = q.poll();

            if(curr.state.equals(target)) return curr.turns;

            for (int i = 0; i < 4; i++) {
                String s1 = curr.state.substring(0, i) +
                            (char)((curr.state.charAt(i) - '0' + 1) % 10 + '0') + 
                            curr.state.substring(i+1);

                String s2 = curr.state.substring(0, i) + 
                            (char)((curr.state.charAt(i) - '0' - 1) % 10 + '0') + 
                            curr.state.substring(i+ 1);

                if(!vis.contains(s1) && !dead.contains(s1))
                {
                    vis.add(s1);
                    q.offer(new lockState(s1, curr.turns + 1));
                }

                if(!vis.contains(s2) && !dead.contains(s2))
                {
                    vis.add(s2);
                    q.offer(new lockState(s2, curr.turns + 1));
                }
                
            }
            
        }
        return -1;
    }
}